File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/go/exercises/practice/connect/connect.go
package connect

import "strings"

func ResultOf(lines []string) (string, error) {
	// Parse grid
	var grid [][]string
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" {
			continue
		}
		grid = append(grid, strings.Fields(trimmed))
	}

	if len(grid) == 0 {
		return "", nil
	}

	// Check O's win (top to bottom)
	oVisited := make(map[[2]int]bool)
	var oQueue [][2]int
	for j, cell := range grid[0] {
		if cell == "O" {
			pos := [2]int{0, j}
			oQueue = append(oQueue, pos)
			oVisited[pos] = true
		}
	}

	for len(oQueue) > 0 {
		current := oQueue[0]
		oQueue = oQueue[1:]
		i, j := current[0], current[1]

		// Check if reached bottom row
		if i == len(grid)-1 {
			return "O", nil
		}

		// Get hexagonal directions based on row parity
		var dirs [][]int
		if i%2 == 0 {  // even row
			dirs = [][]int{{-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, 0}, {1, -1}}
		} else {  // odd row
			dirs = [][]int{{-1, -1}, {-1, 0}, {0, -1}, {0, 1}, {1, -1}, {1, 0}}
		}
		
		for _, dir := range dirs {
			ni, nj := i+dir[0], j+dir[1]
			if ni >= 0 && ni < len(grid) && nj >= 0 && nj < len(grid[ni]) {
				pos := [2]int{ni, nj}
				if !oVisited[pos] && grid[ni][nj] == "O" {
					oVisited[pos] = true
					oQueue = append(oQueue, pos)
				}
			}
		}
	}

	// Check X's win (left to right)
	xVisited := make(map[[2]int]bool)
	var xQueue [][2]int
	for i, row := range grid {
		if len(row) == 0 {
			continue
		}
		if row[0] == "X" {
			pos := [2]int{i, 0}
			xQueue = append(xQueue, pos)
			xVisited[pos] = true
		}
	}

	for len(xQueue) > 0 {
		current := xQueue[0]
		xQueue = xQueue[1:]
		i, j := current[0], current[1]

		// Check if reached rightmost column
		if j == len(grid[i])-1 {
			return "X", nil
		}

		// Get hexagonal directions based on row parity
		var dirs [][]int
		if i%2 == 0 {  // even row
			dirs = [][]int{{-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, 0}, {1, -1}}
		} else {  // odd row
			dirs = [][]int{{-1, -1}, {-1, 0}, {0, -1}, {0, 1}, {1, -1}, {1, 0}}
		}
		
		for _, dir := range dirs {
			ni, nj := i+dir[0], j+dir[1]
			if ni >= 0 && ni < len(grid) && nj >= 0 && nj < len(grid[ni]) {
				pos := [2]int{ni, nj}
				if !xVisited[pos] && grid[ni][nj] == "X" {
					xVisited[pos] = true
					xQueue = append(xQueue, pos)
				}
			}
		}
	}

	// No winner found
	return "", nil
}

File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/python/exercises/practice/connect/connect.py

class ConnectGame:
    def __init__(self, board):
        self.grid = []
        for line in board:
            stripped = line.strip()
            if stripped:
                self.grid.append(list(stripped))
        self.rows = len(self.grid)
        self.cols = len(self.grid[0]) if self.rows else 0

    def get_winner(self):
        if not self.grid:
            return ''
        
        # Check O's vertical connection (top to bottom)
        visited_o = set()
        queue_o = []
        # Start positions are all 'O's in the first row
        for col in range(self.cols):
            if self.grid[0][col] == 'O':
                queue_o.append((0, col))
                visited_o.add((0, col))
        # BFS for O
        for row, col in queue_o:
            if row == self.rows - 1:
                return 'O'
        while queue_o:
            row, col = queue_o.pop(0)
            # Get neighbors based on hexagonal grid parity (even/odd row)
            if row % 2 == 0:
                neighbors = [
                    (row - 1, col),      # up
                    (row - 1, col + 1),  # up-right
                    (row, col - 1),      # left
                    (row, col + 1),      # right
                    (row + 1, col),      # down
                    (row + 1, col + 1),  # down-right
                ]
            else:
                neighbors = [
                    (row - 1, col - 1),  # up-left
                    (row - 1, col),      # up
                    (row, col - 1),      # left
                    (row, col + 1),      # right
                    (row + 1, col - 1),  # down-left
                    (row + 1, col),      # down
                ]
            
            for nr, nc in neighbors:
                if 0 <= nr < self.rows and 0 <= nc < self.cols:
                    if (nr, nc) not in visited_o and self.grid[nr][nc] == 'O':
                        if nr == self.rows - 1:
                            return 'O'
                        visited_o.add((nr, nc))
                        queue_o.append((nr, nc))
        
        # Check X's horizontal connection (left to right)
        visited_x = set()
        queue_x = []
        # Start positions are all 'X's in first column
        for row in range(self.rows):
            if self.grid[row][0] == 'X':
                queue_x.append((row, 0))
                visited_x.add((row, 0))
        # BFS for X
        for row, col in queue_x:
            if col == self.cols - 1:
                return 'X'
        while queue_x:
            row, col = queue_x.pop(0)
            # Get neighbors based on hexagonal grid parity (even/odd row)
            if row % 2 == 0:
                neighbors = [
                    (row - 1, col),      # up
                    (row - 1, col + 1),  # up-right
                    (row, col - 1),      # left
                    (row, col + 1),      # right
                    (row + 1, col),      # down
                    (row + 1, col + 1),  # down-right
                ]
            else:
                neighbors = [
                    (row - 1, col - 1),  # up-left
                    (row - 1, col),      # up
                    (row, col - 1),      # left
                    (row, col + 1),      # right
                    (row + 1, col - 1),  # down-left
                    (row + 1, col),      # down
                ]
            
            for nr, nc in neighbors:
                if 0 <= nr < self.rows and 0 <= nc < self.cols:
                    if (nr, nc) not in visited_x and self.grid[nr][nc] == 'X':
                        if nc == self.cols - 1:
                            return 'X'
                        visited_x.add((nr, nc))
                        queue_x.append((nr, nc))
        
        return ''
